File: projects/engineering.purdue.edu/kak/scriptingwithobjects/swocode/chap15/UDPServer.pl
#!/usr/bin/perl -w
use strict;

### UDPServer.pl

use IO::Socket;                                                      #(A)
use constant MAX_MSG_LEN => 2048;                                    #(B)

my $server_sock = IO::Socket::INET->new( LocalPort => 9000,          #(C)
                                         Reuse => 1,                 #(D)
                                         Proto => 'udp',             #(E)
                                       );
die "No UDP Server Socket: $@\n" unless $server_sock;                #(F)

my ($mesg_in, $mesg_out);                                            #(G)
my $i = 0;
while (1) {                                                          #(H)
    next unless $server_sock->recv( $mesg_in, MAX_MSG_LEN );         #(I)
    $i++;
    my $peerhost = gethostbyaddr( $server_sock->peeraddr, AF_INET )
        || $server_sock->peerhost;                                   #(J)
    my $peerport = $server_sock->peerport;                           #(K)
    my $length = length( $mesg_in );                                 #(L)
    my $client_send_sequence = $mesg_in =~ /^(\d+):/;
    warn "Received $length bytes from [$peerhost, $peerport]\n";     #(M)
    my $lucky_num = int rand 1000;
    $mesg_out = "(Request: $i)" . 
                "(Client Send Sequence Num: $client_send_sequence)" .
                " Your lucky number: $lucky_num";                    #(N)
    $server_sock->send( $mesg_out ) or die "send(): $!\n";           #(O)
}
$server_sock->close;                                                 #(P)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap15/UDPServer.py
#!/usr/bin/env python

### UDPServer.py

import sys, socket, os, signal, re, random                           #(A)

def interrupt_sig_handler( signum, frame ):                          #(B)
    os.kill( os.getpid(), signal.SIGKILL )                           #(C)
signal.signal( signal.SIGINT,  interrupt_sig_handler )               #(D)

ran = random.Random()                                                #(E)

port = 9000                                                          #(F)
MAX_MSG_LEN = 2048                                                   #(G)

try:                                                                 #(H)
    server_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)   #(I)   
    server_sock.bind( ('', port) ) 
except socket.error, (value, message):                               #(K)
    if server_sock:                                                  #(L)
        server_sock.close()                                          #(M)
    print "No server socket: %s" % message                           #(N)
    sys.exit(1)                                                      #(O)

i = 0;                                                               #(P)
while 1:                                                             #(Q)
    mesg_in, client = server_sock.recvfrom( MAX_MSG_LEN )            #(R)
    if not mesg_in: continue                                         #(S)
    i += 1                                                           #(T)
    peerhost, peerport = client                                      #(U)
    length = len( mesg_in )                                          #(V)
    m = re.search( r'^(\d+):', mesg_in )                             #(W)
    if m != None:                                                    #(X)
        client_send_sequence = m.group(1)                            #(Y)
    else:                                                            #(Z)
        client_send_sequence = ''                                    #(a)
    print "Received %s bytes from [%s, %s]" \
                                % (length, peerhost, peerport)       #(b)
    lucky_num = ran.randint( 0, 1000 )                               #(c)
    mesg_out = "[Request: %s] " % i + \
               "[Client Send Sequence Num: %s] " \
                                % client_send_sequence + \
               " Your lucky number: %s" % lucky_num                  #(d)
    server_sock.sendto( mesg_out, client )                           #(e)

server_sock.close()                                                  #(f)
